File: test/samples/sizedstruct/sizedstruct.lua
-- n: the size of the lists
-- lists: a list of list of different sizes
-- strings: a list of strings of different sizes
-- matrices: a list of matrices of different sizes
-- same: a list of list of same sizes
function sized_struct(n, lists, strings, matrices, same)
    -- TODO The is a special case.
end

local n = tonumber(io.read())
local lists = {}
for i = 1, n do
    lists[i] = {}
    lists[i]["size1"] = tonumber(io.read())
    lists[i]["int list"] = {}
    for j in string.gmatch(io.read(), "-?%d+") do
        table.insert(lists[i]["int list"], tonumber(j))
    end
end
local strings = {}
for i = 1, n do
    strings[i] = {}
    strings[i]["size2"] = tonumber(io.read())
    strings[i]["string list"] = io.read()
end
local matrices = {}
for i = 1, 2 do
    matrices[i] = {}
    matrices[i]["size3"] = tonumber(io.read())
    matrices[i]["list list"] = {}
    for j = 1, matrices[i]["size3"] do
        matrices[i]["list list"][j] = {}
        for k in string.gmatch(io.read(), "-?%d+") do
            table.insert(matrices[i]["list list"][j], tonumber(k))
        end
    end
end
local same = {}
for i = 1, n do
    same[i] = {}
    same[i]["size4"] = tonumber(io.read())
    same[i]["int list n"] = {}
    for j in string.gmatch(io.read(), "-?%d+") do
        table.insert(same[i]["int list n"], tonumber(j))
    end
end

sized_struct(n, lists, strings, matrices, same)

File: test/samples/sizedstruct/sizedstruct.hs
import Control.Monad (replicateM)

-- | contains a list
data List = List
  { size1   :: Int    -- ^ the list's size
  , intList :: [Int]  -- ^ the integer list
  }

-- | contains a string
data String' = String'
  { size2      :: Int     -- ^ the list's size
  , stringList :: String  -- ^ the string list
  }

-- | contains a matrix
data Matrix = Matrix
  { size3    :: Int      -- ^ the list's size
  , listList :: [[Int]]  -- ^ the list list
  }

-- | this is not a 'sized struct', but a regular one!
data NotASizedStruct = NotASizedStruct
  { size4    :: Int    -- ^ not the list's size
  , intListN :: [Int]  -- ^ the integer list
  }

sizedStruct :: Int                -- ^ the size of the lists
            -> [List]             -- ^ a list of list of different sizes
            -> [String']          -- ^ a list of strings of different sizes
            -> [Matrix]           -- ^ a list of matrices of different sizes
            -> [NotASizedStruct]  -- ^ a list of list of same sizes
            -> String             -- ^ TODO
-- The is a special case.
sizedStruct n lists strings matrices same = "TODO"

main :: IO ()
main = do
  n <- fmap read getLine
  lists <- replicateM n readList
  strings <- replicateM n readString'
  matrices <- replicateM 2 readMatrix
  same <- replicateM n readNotASizedStruct
  putStrLn $ sizedStruct n lists strings matrices same
  where
    readList = List <$> fmap read getLine <*> fmap (map read . words) getLine
    readString' = String' <$> fmap read getLine <*> getLine
    readMatrix = fmap read getLine >>= \a -> Matrix a <$> (replicateM a $ fmap (map read . words) getLine)
    readNotASizedStruct = NotASizedStruct <$> fmap read getLine <*> fmap (map read . words) getLine
